spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala

```diff
@@ -1143,66 +1147,64 @@ trait SnapshotManagement { self: DeltaLog =>
+    getSnapshotAt(version, lastCheckpointHint, lastCheckpointProvider = None)
+  }
+
   /**
-   * Get the snapshot at `version` using the given `lastCheckpointProvider` hint
+   * Get the snapshot at `version` using the given `lastCheckpointProvider` or `lastCheckpointHint`
    * as the listing hint.
    */
   private[delta] def getSnapshotAt(
       version: Long,
-      lastCheckpointProvider: CheckpointProvider): Snapshot = {
+      lastCheckpointHint: Option[CheckpointInstance],
+      lastCheckpointProvider: Option[CheckpointProvider]): Snapshot = {
+
     // See if the version currently cached on the cluster satisfies the requirement
-    val current = unsafeVolatileSnapshot
-    if (current.version == version) {
-      return current
+    val currentSnapshot = unsafeVolatileSnapshot
+    val upperBoundSnapshot = if (currentSnapshot.version >= version) {
+      // current snapshot is already newer than what we are looking for. so it could be used as
+      // upper bound.
+      currentSnapshot
+    } else {
+      val latestSnapshot = update()
+      if (latestSnapshot.version < version) {
+        throwNonExistentVersionError(version)
+      }
+      latestSnapshot
     }
-    if (lastCheckpointProvider.version > version) {
-      // if the provided lastCheckpointProvider's version is greater than the snapshot that we are
-      // trying to create => we can't use the provider.
-      // fallback to the other overload.
-      return getSnapshotAt(version)
+    if (upperBoundSnapshot.version == version) {
+      return upperBoundSnapshot
     }
-    val segment = createLogSegment(
-      versionToLoad = Some(version),
-      oldCheckpointProviderOpt = Some(lastCheckpointProvider)
-    ).getOrElse {
-      // We can't return InitialSnapshot because our caller asked for a specific snapshot version.
-      throw DeltaErrors.emptyDirectoryException(logPath.toString)
-    }
-    createSnapshot(
-      initSegment = segment,
-      commitStoreOpt = None,
-      checksumOpt = None)
-  }
 
-  /** Get the snapshot at `version`. */
-  def getSnapshotAt(
-      version: Long,
-      lastCheckpointHint: Option[CheckpointInstance] = None): Snapshot = {
-    // See if the version currently cached on the cluster satisfies the requirement
-    val current = unsafeVolatileSnapshot
-    if (current.version == version) {
-      return current
+    val (lastCheckpointInfoOpt, lastCheckpointProviderOpt) = lastCheckpointProvider match {
+      // NOTE: We must ignore any hint whose version is higher than the requested version.
+      case Some(checkpointProvider) if checkpointProvider.version <= version =>
+        // Prefer the last checkpoint provider hint, because it doesn't require any I/O to use.
+        None -> Some(checkpointProvider)
+      case _ =>
+        val lastCheckpointInfoForListing = lastCheckpointHint
+            .filter(_.version <= version)
+            .orElse(findLastCompleteCheckpointBefore(version))
```

Comment:
This does backward listing in advance, so `createLogSegment` won't have to list from 0?
(probably worth a comment, a bit subtle)

Reply:
Yes - Adding doc comment in a followup PR.